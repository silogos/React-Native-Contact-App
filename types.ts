export type Contact = {
  id: string;
  firstName: string;
  lastName: string;
  age?: number;
  photo?: string;
};