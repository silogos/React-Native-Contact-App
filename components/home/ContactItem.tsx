import { Image } from "expo-image";
import { Link } from "expo-router";
import { useEffect, useState } from "react";
import { Pressable, StyleSheet, View } from "react-native";

import { ThemedText } from "@/components/ThemedText";
import { Contact } from "@/types";

export type ContactItemProps = {
  contact: Contact;
};

export function ContactItem({
  contact: { id, firstName, lastName, photo },
}: ContactItemProps) {
  const [photoUrl, setPhotoUrl] = useState<string>();

  useEffect(() => {
    if (photo) setPhotoUrl(photo);
  }, [photo]);

  return (
    <Link href={`/detail/${id}`} asChild>
      <Pressable style={styles.container}>
        <View style={styles.image}>
          {photoUrl ? (
            <Image
              style={{ flex: 1, width: "100%" }}
              source={photoUrl}
              contentFit="cover"
              transition={1000}
              onError={() => setPhotoUrl(undefined)}
            />
          ) : (
            <ThemedText type="defaultSemiBold" style={{ color: "#FFF" }}>
              {firstName[0].toUpperCase() + lastName[0].toUpperCase()}
            </ThemedText>
          )}
        </View>
        <ThemedText
          colorName="foreground"
          type="defaultSemiBold"
          style={styles.name}
        >
          {firstName + " " + lastName}
        </ThemedText>
      </Pressable>
    </Link>
  );
}

const styles = StyleSheet.create({
  container: {
    flexDirection: "row",
    alignItems: "center",
    paddingHorizontal: 18,
    paddingVertical: 10,
    gap: 5,
  },
  image: {
    width: 50,
    height: 50,
    backgroundColor: "#9CA0AE",
    borderRadius: 25,
    overflow: "hidden",
    justifyContent: "center",
    alignItems: "center",
  },
  name: {
    flex: 1,
    padding: 10,
  },
});
