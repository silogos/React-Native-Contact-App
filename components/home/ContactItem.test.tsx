import React from "react";
import { render, waitFor } from "@testing-library/react-native";

import { ContactItem } from "./ContactItem";
import { Contact } from "@/types";

describe("Nav", () => {
  it("renders correctly", () => {
    const contact: Contact = {
      id: "",
      firstName: "D",
      lastName: "D",
      photo: "asdasdas",
    };
    const elm = render(<ContactItem contact={contact} />).toJSON();

    expect(elm).toMatchSnapshot();
  });

  it("Render with photo undefined", async () => {
    const contact: Contact = {
      id: "",
      firstName: "D",
      lastName: "D",
      photo: undefined,
    };
    const { firstName, lastName } = contact;
    const { getByText } = render(<ContactItem contact={contact} />);
    const titleElement = getByText(
      `${firstName[0].toUpperCase()} ${lastName[0].toUpperCase()}`
    );

    expect(titleElement).toBeTruthy();
  });

  it("Render with photo defined", async () => {
    const contact: Contact = {
      id: "",
      firstName: "D",
      lastName: "D",
      photo: "asdasdas",
    };

    const { getByTestId } = render(<ContactItem contact={contact} />);

    expect(getByTestId("contact-photo")).toBeTruthy();
  });
});
