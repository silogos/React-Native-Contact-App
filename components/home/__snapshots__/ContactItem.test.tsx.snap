// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Nav renders correctly 1`] = `
<View
  accessibilityState={
    {
      "busy": undefined,
      "checked": undefined,
      "disabled": undefined,
      "expanded": undefined,
      "selected": undefined,
    }
  }
  accessibilityValue={
    {
      "max": undefined,
      "min": undefined,
      "now": undefined,
      "text": undefined,
    }
  }
  accessible={true}
  collapsable={false}
  focusable={true}
  href="/detail/"
  onBlur={[Function]}
  onClick={[Function]}
  onFocus={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  role="link"
  style={
    {
      "alignItems": "center",
      "flexDirection": "row",
      "gap": 5,
      "paddingHorizontal": 18,
      "paddingVertical": 10,
    }
  }
>
  <View
    style={
      {
        "alignItems": "center",
        "backgroundColor": "#9CA0AE",
        "borderRadius": 25,
        "height": 50,
        "justifyContent": "center",
        "overflow": "hidden",
        "width": 50,
      }
    }
  >
    <ViewManagerAdapter_ExpoImage
      contentFit="cover"
      contentPosition={
        {
          "left": "50%",
          "top": "50%",
        }
      }
      flex={1}
      nativeViewRef={"[React.ref]"}
      onError={[Function]}
      onLoad={[Function]}
      onLoadStart={[Function]}
      onProgress={[Function]}
      placeholder={[]}
      source={
        [
          {
            "uri": "asdasdas",
          },
        ]
      }
      style={
        {
          "flex": 1,
          "width": "100%",
        }
      }
      testID="contact-photo"
      transition={
        {
          "duration": 1000,
        }
      }
      width="100%"
    />
  </View>
  <Text
    style={
      [
        {
          "color": "#09090b",
        },
        undefined,
        undefined,
        {
          "fontSize": 16,
          "fontWeight": "600",
        },
        undefined,
        undefined,
        {
          "flex": 1,
          "padding": 10,
        },
      ]
    }
  >
    D D
  </Text>
</View>
`;
